### PR TITLE
Implement Frequencies + other utility functions

### PR DESCRIPTION
**Pull request by kedartatwawadi**
I didn't get time to write about arithmetic coding, so that PR is still in draft version (i.e. not ready to review) 
I think it might be beneficial to merge some of the utils until then.

### CODE REVIEW

File: core/prob_dist.py [hunk 1]
```diff
@@ -100,3 +100,93 @@ def test_prob_creation_and_validation(self):
 
         # check if this works
         _ = ProbabilityDist(dist)
+
+
+def get_mean_log_prob(prob_dist: ProbabilityDist, data_block) -> float:
+    """computes the average log_probability of the input data_block given the probability distribution
+    prob_dist. This acts as a fundamental limit on what any compressor designed for distribution
```
Comment (shubhamchandak94) [inline core/prob_dist.py:107]: "This acts as a fundamental limit on what any ..." - I would make this weaker statement. It's intuitively correct for most standard entropy coders "designed for distribution prob_dist", but one can always created weird codes that give small codewords to low probability symbols to beat this bound for certain sequences

File: core/prob_dist.py [hunk 2]
```diff
@@ -100,3 +100,93 @@ def test_prob_creation_and_validation(self):
...
+
+class Frequencies:
+    """
+    Wrapper around a frequency dict
+    NOTE: Frequencies is a typical way to represent probability distributions using integers
+    """
+
+    def __init__(self, freq_dict=None):
+
+        # NOTE: We use the fact that since python 3.6, dictionaries in python are
```
Comment (shubhamchandak94) [inline core/prob_dist.py:131]: where do we use this fact? Is it for the lists?

File: core/prob_dist.py [hunk 3]
```diff
@@ -100,3 +100,93 @@ def test_prob_creation_and_validation(self):
...
+
+    @property
+    def alphabet(self):
+        return list(self.freq_dict)
+
+    @property
+    def freq_list(self):
+        return [self.freq_dict[s] for s in self.alphabet]
+
+    @property
```
Comment (shubhamchandak94) [inline core/prob_dist.py:150]: TODO - do the cached thing

File: core/prob_dist.py (hunk 2 above)
Comment (shubhamchandak94) [reply core/prob_dist.py:131]: I guess also for cumulative

File: utils/bitarray_utils.py [hunk 4]
```diff
@@ -4,8 +4,16 @@
 from typing import Tuple
 
 
-def get_bit_width(size) -> int:
-    return int(np.ceil(np.log2(size)))
+def get_bit_width(x) -> int:
```
Comment (shubhamchandak94) [inline utils/bitarray_utils.py:7]: TODO - add test for this, checking for boundary values like powers of 2, powers of 2 minus 1 (looks correct to me but I have made mistakes in the past, so better to add simple tests)
also add assertion for positivity
do we want to support x=0

File: utils/bitarray_utils.py [hunk 5]
```diff
@@ -18,7 +26,7 @@ def uint_to_bitarray(x: int, bit_width=None) -> BitArray:
     converts an unsigned into to bits.
     if bit_width is provided then data is converted accordingly
     """
-    return int2ba(x, length=bit_width)
+    return int2ba(int(x), length=bit_width)
```
Comment (shubhamchandak94) [inline utils/bitarray_utils.py:29]: why converting to int? Is it better to give an error than to silently convert to int

File: utils/test_utils.py [hunk 6]
```diff
@@ -69,7 +69,8 @@ def try_lossless_compression(
     output_bits_block = encoder.encode_block(data_block)
 
     # test decode
-    decoded_block, _ = decoder.decode_block(output_bits_block)
+    decoded_block, num_bits_consumed = decoder.decode_block(output_bits_block)
+    assert num_bits_consumed == len(output_bits_block)
```
Comment (shubhamchandak94) [inline utils/test_utils.py:73]: Nice. Wonder how we missed this before

File: core/prob_dist.py (hunk 1 above)
Comment (kedartatwawadi) [reply core/prob_dist.py:107]: rewrote as 
```
NOTE: although not a funcdamental limit, this is roughly what a compressor designed for the given distribution prob_dist will aim to achieve"
```

File: core/prob_dist.py [hunk 7]
```diff
@@ -100,3 +100,93 @@ def test_prob_creation_and_validation(self):
...
+        prob_dist (ProbabilityDist): specified probability distribution used to compute log_probability
+        data_block (DataBlock): input for which avg log probability needs to be computed
+    """
+
+    # get avg log probability for the input
+    log_prob = 0
+    for s in data_block.data_list:
+        log_prob += prob_dist.log_probability(s)
+    avg_log_prob = log_prob / data_block.size
+    return avg_log_prob
```
[the remark this answers is not in the document]
Comment (kedartatwawadi) [reply core/prob_dist.py:120]: done

File: core/prob_dist.py (hunk 2 above)
Comment (kedartatwawadi) [reply core/prob_dist.py:131]: yes exactly

File: core/prob_dist.py (hunk 3 above)
Comment (kedartatwawadi) [reply core/prob_dist.py:150]: done

File: utils/bitarray_utils.py (hunk 4 above)
Comment (kedartatwawadi) [reply utils/bitarray_utils.py:7]: good point, I think there is indeed a mistake here! 
if x = 4, this returns 2, which is not enough :-|

Comment (kedartatwawadi) [reply utils/bitarray_utils.py:7]: fixed + added tests

File: utils/bitarray_utils.py (hunk 5 above)
Comment (kedartatwawadi) [reply utils/bitarray_utils.py:29]: yeah, not needed it looks like

Comment (kedartatwawadi) [reply utils/bitarray_utils.py:29]: added an assert

File: utils/test_utils.py (hunk 6 above)
Comment (kedartatwawadi) [reply utils/test_utils.py:73]: actually want to improve this test..  by optionally adding some extra bits to encoder output. This is to simulate the scenario where there are additional bits beyond the bitstream encoded. Some compressors (like range coder, AEC) support this .

Comment (kedartatwawadi) [reply utils/test_utils.py]: @shubhamchandak94: take a re-look

**Review by shubhamchandak94 — APPROVED** [review]

File: core/prob_dist.py (hunk 1 above)
Comment (tpulkit) [reply core/prob_dist.py:113]: I don't see the change in the latest committed files.

File: core/prob_dist.py (hunk 2 above)
Comment (tpulkit) [reply core/prob_dist.py:137]: Why do you need it to be OrderedDict? We don't really care order in which symbols were added to the dict, correct?

File: core/prob_dist.py (hunk 1 above)
Comment (kedartatwawadi) [reply core/prob_dist.py:113]: I did update it a bit, but still kept the "fundamental limit" business.. ll update

File: core/prob_dist.py (hunk 2 above)
Comment (kedartatwawadi) [reply core/prob_dist.py:137]: We do need ordering. for example when we return the `freq_list` or the `cumulative_freq_list`.. we assume that the first frequency in the list corresponds to  the first symbol (for example).

Comment (shubhamchandak94) [reply core/prob_dist.py:137]: I think the important part in ordering is the consistency. The particular ordering according to insertion order is probably not that important because we will almost always insert all symbols at the beginning and later only update frequencies

Comment (tpulkit) [reply core/prob_dist.py:137]: Got it. Thanks.

File: core/prob_dist.py (hunk 1 above)
Comment (tpulkit) [reply core/prob_dist.py]: English can be improved, but I will merge it for now. I think I will have a look at all the comments before class just to see the language makes sense. I think many of my comments are also not good grammatically.

**Review by tpulkit — COMMENTED** [review]
Looks good. Merging.

**Review by tpulkit — APPROVED** [review]
